sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1715,31 +1715,68 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
       // Filter the plan by applying left semi and left anti joins.
       withSubquery.foldLeft(newFilter) {
         case (p, PredicateSubquery(sub, conditions, _, _)) =>
-          Join(p, sub, LeftSemi, conditions.reduceOption(And))
+          if (!conditions.exists(PredicateSubquery.hasPredicateSubquery)) {
```

Comment:
So the code for these rewrites is (almost) the same for all three cases. Lets move this into a helper method or move it into a separate a loop before this one (I prefer the latter).